### DISCUSSION TITLE
Which ports used for direct connection

### DISCUSSION THREAD

**Discussion by glynj-bsquare** [paullouisageneau/libdatachannel#470] [Q&A] [answered]
Hi,
If the two devices connecting are on the same network (e.g. in a test lab), then the system decides to use a direct UDP (or TCP) connection between the devices as the best route.
One device listens on the port, then the other device connects to it.
This means that any firewall on the listening device needs to be configured for the possible port range that could be used, otherwise the OS drops the connection opened for the data channel when the other end tries to connect.
Is there a way to limit or define which ports can be used?
If not, what does the code currently use as the range to choose from?
Thanks,
Glyn.

**Comment by paullouisageneau** [comment, marked as answer]
Hi,
Even if some WebRTC agents can use a local TCP connection where one device listens and one device connects (Firefox supports it, libdatachannel supports it only with libnice backend), that's not how WebRTC connections are open in general. WebRTC primarily uses UDP and relies on ICE to traverse both NATs and firewalls.

For firewalls, the idea is that the connection is negotiated and open from both sides at the same time, meaning UDP datagrams are send from each side to establish the connection. Therefore, firewalls on each side will consider the remote connection attempt as an answer to the local connection attempt.

By default any available port above 1024 may be used, but there are settings `portRangeBegin` and `portRangeEnd` in `rtc::Configuration` (see [configuration.hpp](https://github.com/paullouisageneau/libdatachannel/blob/86c0d2687ccf40e9c1b9ad07d97a6aa811571716/include/rtc/configuration.hpp#L88)) to limit the UDP port range, in case it is necessary.

**Reply by glynj-bsquare** [reply]
Thanks for the quick reply.
